rspirv/lift/storage.rs

```diff
@@ -0,0 +1,59 @@
+use crate::{
+    sr::{
+        storage::{Storage, Token},
+    },
+};
+
+use std::{
+    borrow::Borrow,
+    collections::hash_map::{Entry, VacantEntry, HashMap},
+    hash::BuildHasherDefault,
+};
+
+use fxhash::FxHasher;
+
+
+pub struct LiftStorage<T, L = Token<T>> {
+    values: Storage<T>,
+    lookup: HashMap<spirv::Word, L, BuildHasherDefault<FxHasher>>,
+}
+
+impl<T, L: Borrow<Token<T>>> LiftStorage<T, L> {
+    pub(in crate::lift) fn new() -> Self {
+        LiftStorage {
+            values: Storage::new(),
+            lookup: HashMap::default(),
+        }
+    }
+
+    pub(in crate::lift) fn unwrap(self) -> Storage<T> {
+        self.values
+    }
+
+    pub(in crate::lift) fn lookup(
+        &self, id: spirv::Word
+    ) -> (&T, &L) {
+        let info = &self.lookup[&id];
+        (&self.values[*info.borrow()], info)
+    }
+
+    pub(in crate::lift) fn append(
+        &mut self, id: spirv::Word, value: T
+    ) -> (Token<T>, VacantEntry<spirv::Word, L>) {
+        let token = self.values.append(value);
+        match self.lookup.entry(id) {
+            Entry::Occupied(_) => panic!("Id {:?} is already used", id),
+            Entry::Vacant(e) => (token, e),
+        }
+    }
+}
+
+impl<T> LiftStorage<T> {
```

Comment:
It has been a while and this syntax is not quite familiar to me anymore. What's the difference of this `impl` compared to the above one?

Reply:
The one above is generic across `L`. This block omits `L` thus it's for the default `L = Token<T>` only, which is a narrower subset of all.